tooling/nargo_fmt/src/formatter/chunks.rs

```diff
@@ -0,0 +1,776 @@
+//!
+//! However, some changes were introduces to handle comments and other particularities of Noir.
+use super::Formatter;
+
+/// A text chunk. It precomputes the text width and wether it has newlines.
+#[derive(Debug)]
+pub(crate) struct TextChunk {
+    pub(crate) string: String,
+    pub(crate) width: usize,
+    pub(crate) has_newlines: bool,
+}
+
+impl TextChunk {
+    pub(crate) fn new(string: String) -> Self {
+        TextChunk {
+            width: string.chars().count(),
+            has_newlines: string.chars().any(|char| char == '\n'),
+            string,
+        }
+    }
+}
+
+/// A chunk can either be text or a directive that instructs the formatter to do something
+/// (for example: increase or decrease the current indentation).
+#[derive(Debug)]
+pub(crate) enum Chunk {
+    /// A text chunk. It might contain leading comments.
+    Text(TextChunk),
+    /// A trailing comma that's only written if we decide to format chunks in multiple lines
+    /// (for example for a call we'll add a trailing comma to the last argument).
+    TrailingComma,
+    /// A trailing comment (happens at the end of a line, and always after something else have been written).
+    TrailingComment(TextChunk),
+    /// A leading comment. Happens at the beginning of a line.
+    LeadingComment(TextChunk),
+    /// A group of chunks.
+    Group(ChunkGroup),
+    /// Write a line (or two) if we decide to format chunks in multiple lines, otherwise do nothing.
+    Line { two: bool },
+    /// Writes a space if we can write a group in one line, otherwise writes a line.
+    /// However, a space might be written if `one_chunk_per_line` of a Chunks object is set to false.
+    SpaceOrLine,
+    /// Command to increase the current indentation.
+    IncreaseIndentation,
+    /// Command to decrease the current indentation.
+    DecreaseIndentation,
+    /// Push the current indentation to the indentation stack.
+    PushIndentation,
+    /// Set the current indentation by popping it from the indentation stack.
+    PopIndentation,
+}
+
+impl Chunk {
+    pub(crate) fn width(&self) -> usize {
+        match self {
+            Chunk::Text(chunk) | Chunk::TrailingComment(chunk) | Chunk::LeadingComment(chunk) => {
+                chunk.width
+            }
+            Chunk::Group(group) => group.width(),
+            Chunk::SpaceOrLine => 1,
+            Chunk::Line { .. }
+            | Chunk::IncreaseIndentation
+            | Chunk::DecreaseIndentation
+            | Chunk::TrailingComma
+            | Chunk::PushIndentation
+            | Chunk::PopIndentation => 0,
+        }
+    }
+
+    /// Computes the width of this chunk considering it's inside an ExpressionList.
+    /// The only thing that changes here compared to `width` is that a LambdaAsLastExpressionInList's
+    /// width is considered to be only the first line, so we can avoid splitting the entire call
+    /// arguments into separate lins.
```

Comment:
```suggestion
    /// arguments into separate lines.
```
On a different note, perhaps an example of the `LambdaAsLastExpressionInList` would be helpful.

Reply:
Good idea, I'll add an example soon.